File: unit_tests/test_knight.py
import unittest
from board import Board
from constants import PieceType, Color, Direction
from move import Move


class TestSuite_KnightMoves(unittest.TestCase):
    def test_knight_moves_1(self):
        """knight's moves on an empty board"""

        b = Board()
        knight = b.place_piece('a1', PieceType.KNIGHT, Color.WHITE)
        actualMoves = knight.calc_potential_moves()
        expectedMoves = ['Na1-c2', 'Na1-b3']
        self.assertEqual(set(map(str, actualMoves)), set(expectedMoves))

        b = Board()
        knight = b.place_piece('e4', PieceType.KNIGHT, Color.WHITE)
        actualMoves = knight.calc_potential_moves()
        expectedMoves = ['Ne4-d6', 'Ne4-d2', 'Ne4-c3', 'Ne4-c5', 'Ne4-f2', 'Ne4-f6', 'Ne4-g3', 'Ne4-g5', ]
        self.assertEqual(set(map(str, actualMoves)), set(expectedMoves))

    def test_knight_moves_2(self):
        """knight's moves on a non-empty board"""

        b = Board()
        knight = b.place_piece('c1', PieceType.KNIGHT, Color.WHITE)
        b.place_piece('d3', PieceType.PAWN, Color.WHITE)
        b.place_piece('b3', PieceType.PAWN, Color.BLACK)
        actualMoves = knight.calc_potential_moves()
        expectedMoves = ['Nc1-a2', 'Nc1-e2', 'Nc1-b3']
        self.assertEqual(set(map(str, actualMoves)), set(expectedMoves))

    def test_knight_moves_pinned(self):
        """knight's moves while pinned"""

        b = Board()
        knight = b.place_piece('c1', PieceType.KNIGHT, Color.WHITE)
        actualMoves = knight.calc_potential_moves_pinned(Direction.UP)
        expectedMoves = []
        self.assertEqual(set(map(str, actualMoves)), set(expectedMoves))

        actualMoves = knight.calc_potential_moves_pinned(Direction.LEFT)
        self.assertEqual(set(map(str, actualMoves)), set(expectedMoves))

        actualMoves = knight.calc_potential_moves_pinned(Direction.DOWN_RIGHT)
        self.assertEqual(set(map(str, actualMoves)), set(expectedMoves))

    def test_my(self):
        b = Board()
        wk = b.place_piece('h1', PieceType.KING, Color.WHITE)
        wk.update_attacked_squares()

        bk = b.place_piece('h3', PieceType.KING, Color.BLACK)
        bk.update_attacked_squares()

        wn = b.place_piece('f1', PieceType.KNIGHT, Color.WHITE)
        wn.recalculate()

        moves = b.get_all_legal_moves()
        m = Move(wn, b.get_square_by_name('g3'))
        b.perform_move(m)

        print('done')

File: c_engine/test/test_knight.cpp
#include "constants.h"
#include "knight.h"
#include "utest.h"
#include "move.h"

#define private public

#include "board.h"

TestSuite create_test_suite_knight() {
    TestSuite testSuite("Knight");

    testSuite.add_test("Knight construction", []() {
        Board b;
        auto knight = dynamic_cast<Knight*>(b.place_piece(PieceType::KNIGHT, Color::WHITE, "a1"));

        assert_equal(PieceType::KNIGHT, knight->_kind);
        assert_equal(Color::WHITE, knight->_color);
        assert_equal("a1", knight->_square->get_name());
        assert_equal("N", knight->_name);
        assert_true(knight->_isLight);
    });
    
    testSuite.add_test("Knight move generation 1", []() {
        Board b;
        auto knight = dynamic_cast<Knight*>(b.place_piece(PieceType::KNIGHT, Color::WHITE, "a1"));
        knight->recalculate();

        auto moves = b.squares_to_moves(knight->get_potential_squares(), knight);

        assert_equal(2U, moves.size());
        for (auto move : {"Na1-b3", "Na1-c2"})
            assert_vector_contains(moves, move);
    });

    testSuite.add_test("Knight move generation 2", []() {
        Board b;
        auto knight = dynamic_cast<Knight*>(b.place_piece(PieceType::KNIGHT, Color::WHITE, "d4"));
        knight->recalculate();

        auto moves = b.squares_to_moves(knight->get_potential_squares(), knight);

        assert_equal(8U, moves.size());
        for (auto move : {"Nd4-b3", "Nd4-b5", "Nd4-c2", "Nd4-c6", "Nd4-e2", "Nd4-e6", "Nd4-f3", "Nd4-f5"})
            assert_vector_contains(moves, move);
    });

    return testSuite;
}